In [1]:
from datetime import date
import polars as pl
from flights.evaluation.evaluation import main
test_date = date(2026,3,1)
results, df_stats = main(test_src="adsbx", gold_src="sfdps", test_dates = test_date)
df = results[0]

Loading airports from /Volumes/T2-SSD/planequery/data/raw/ourairports/airports.csv
Stats for adsbx:sfdps: {'true_positive': 6244, 'false_positive': 31, 'false_negative': 484, 'precision': 0.9950597609561753, 'recall': 0.9280618311533888, 'f1': 0.9603937552872414, 'takeoff_airport_ident_match_pct': 0.9985586162716208, 'landing_airport_ident_match_pct': 0.9990390775144138, 'airport_ident_match_pct': 0.9975976937860346}
Stats CSV saved to /Volumes/T2-SSD/planequery/data/experiments/comparison/2026-07-15_07-32_ea042da_flights_evaluation.csv


In [3]:
df.filter(pl.col("icao") == "a004e9")

icao,callsign,registration,takeoff_time,takeoff_airport_ident,landing_time,landing_airport_ident,pia,ladd,military,interesting,first_message_time,first_lat,first_lon,last_message_time,last_lat,last_lon,icao_df1,callsign_df1,registration_df1,takeoff_time_df1,takeoff_airport_ident_df1,landing_time_df1,landing_airport_ident_df1,pia_df1,ladd_df1,military_df1,interesting_df1,aircraft_type,owner,aircraft_description,category,first_message_time_df1,first_lat_df1,first_lon_df1,last_message_time_df1,last_lat_df1,last_lon_df1,discard_df1,takeoff_airport_lat_df1,takeoff_airport_lon_df1,landing_airport_lat_df1,landing_airport_lon_df1,match_status,same_takeoff_airport_ident,same_landing_airport_ident,same_airport_ident,first_message_time_diff,last_message_time_diff,first_message_distance_to_takeoff_airport_km,last_message_distance_to_landing_airport_km
str,str,str,datetime[ms],str,datetime[ms],str,bool,bool,bool,bool,datetime[ms],f64,f64,datetime[ms],f64,f64,str,str,str,datetime[ms],str,datetime[ms],str,bool,bool,bool,bool,str,str,str,str,datetime[ms],f64,f64,datetime[ms],f64,f64,bool,f64,f64,f64,f64,str,bool,bool,bool,duration[ms],duration[ms],f64,f64
"""a004e9""",null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""a004e9""","""N100CC""","""N100CC""",2026-03-01 17:55:00,"""KFCM""",2026-03-01 18:50:00,"""KDSM""",false,true,false,false,"""C425""","""SENECA TANK""","""CESSNA 425 Corsair""","""A1""",null,null,null,null,null,null,false,44.827202,-93.4571,41.534027,-93.656719,"""df1_only""",false,false,false,null,null,null,null


In [6]:
COLUMNS = ["icao", "takeoff_time", "landing_time", "takeoff_airport_ident", "landing_airport_ident", "takeoff_time_df1", "landing_time_df1", "takeoff_airport_ident_df1", "landing_airport_ident_df1", "match_status"]
df.filter(pl.col("icao") == "a003a8").select(COLUMNS)

icao,takeoff_time,landing_time,takeoff_airport_ident,landing_airport_ident,takeoff_time_df1,landing_time_df1,takeoff_airport_ident_df1,landing_airport_ident_df1,match_status
str,datetime[ms],datetime[ms],str,str,datetime[ms],datetime[ms],str,str,str
"""a003a8""",null,null,null,null,2026-03-01 18:08:00,2026-03-01 18:59:00,"""KBCT""","""MYEN""","""df1_only"""
"""a003a8""",null,null,null,null,2026-03-01 19:54:00,2026-03-01 20:52:00,"""MYEN""","""KFXE""","""df1_only"""
"""a003a8""",2026-03-01 18:08:13,2026-03-01 20:52:44,"""KBCT""","""KFXE""",null,null,null,null,"""df0_only"""


In [ ]:
# a003a8 wasen't in data before now it may be with adsbx used for flights matching. 

In [4]:
df.filter(pl.col("match_status") == "df0_only")

icao,callsign,registration,takeoff_time,takeoff_airport_ident,landing_time,landing_airport_ident,pia,ladd,military,interesting,first_message_time,first_lat,first_lon,last_message_time,last_lat,last_lon,icao_df1,callsign_df1,registration_df1,takeoff_time_df1,takeoff_airport_ident_df1,landing_time_df1,landing_airport_ident_df1,pia_df1,ladd_df1,military_df1,interesting_df1,aircraft_type,owner,aircraft_description,category,first_message_time_df1,first_lat_df1,first_lon_df1,last_message_time_df1,last_lat_df1,last_lon_df1,discard_df1,takeoff_airport_lat_df1,takeoff_airport_lon_df1,landing_airport_lat_df1,landing_airport_lon_df1,match_status,same_takeoff_airport_ident,same_landing_airport_ident,same_airport_ident,first_message_time_diff,last_message_time_diff,first_message_distance_to_takeoff_airport_km,last_message_distance_to_landing_airport_km
str,str,str,datetime[ms],str,datetime[ms],str,bool,bool,bool,bool,datetime[ms],f64,f64,datetime[ms],f64,f64,str,str,str,datetime[ms],str,datetime[ms],str,bool,bool,bool,bool,str,str,str,str,datetime[ms],f64,f64,datetime[ms],f64,f64,bool,f64,f64,f64,f64,str,bool,bool,bool,duration[ms],duration[ms],f64,f64
"""a003a8""",null,"""N10PH""",2026-03-01 18:08:13,"""KBCT""",2026-03-01 20:52:44,"""KFXE""",false,true,false,false,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""df0_only""",false,false,false,null,null,null,null
"""a09d74""",null,"""N139JM""",2026-03-01 18:00:13,"""KADS""",2026-03-01 23:27:48,"""KBTR""",false,false,false,false,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""df0_only""",false,false,false,null,null,null,null
"""a0af5a""",null,"""N143RP""",2026-03-01 17:20:29,"""KTHV""",2026-03-01 20:52:34,"""KAPF""",false,true,false,false,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""df0_only""",false,false,false,null,null,null,null
"""a1a4a7""",null,"""N205AS""",2026-03-01 15:40:08,"""KBHM""",2026-03-01 17:19:16,"""KNEW""",false,false,false,false,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""df0_only""",false,false,false,null,null,null,null
"""a1a4a7""",null,"""N205AS""",2026-03-01 17:23:47,"""KNEW""",2026-03-01 19:18:47,"""KBHM""",false,false,false,false,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""df0_only""",false,false,false,null,null,null,null
…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…
"""aa69e0""",null,"""N770JS""",2026-03-01 16:00:11,"""KMDW""",2026-03-01 20:06:18,"""KBCT""",false,false,false,false,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""df0_only""",false,false,false,null,null,null,null
"""aa9d1a""",null,"""N783QS""",2026-03-01 15:40:47,"""KBED""",2026-03-01 21:34:08,"""TJIG""",false,false,false,false,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""df0_only""",false,false,false,null,null,null,null
"""aae0db""",null,"""N80PT""",2026-03-01 14:03:33,"""KNEW""",2026-03-01 20:22:50,"""KBTV""",false,true,false,false,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""df0_only""",false,false,false,null,null,null,null


In [7]:
import polars as pl
df.filter(pia_or_american_ladd_expr()).filter(pl.col("match_status") == "df0_only")

icao,callsign,registration,takeoff_time,takeoff_airport_ident,landing_time,landing_airport_ident,pia,ladd,military,interesting,first_message_time,first_lat,first_lon,last_message_time,last_lat,last_lon,icao_df1,callsign_df1,registration_df1,takeoff_time_df1,takeoff_airport_ident_df1,landing_time_df1,landing_airport_ident_df1,pia_df1,ladd_df1,military_df1,interesting_df1,aircraft_type,owner,aircraft_description,category,first_message_time_df1,first_lat_df1,first_lon_df1,last_message_time_df1,last_lat_df1,last_lon_df1,discard_df1,takeoff_airport_lat_df1,takeoff_airport_lon_df1,landing_airport_lat_df1,landing_airport_lon_df1,match_status,same_takeoff_airport_ident,same_landing_airport_ident,same_airport_ident,first_message_time_diff,last_message_time_diff,first_message_distance_to_takeoff_airport_km,last_message_distance_to_landing_airport_km
str,str,str,datetime[ms],str,datetime[ms],str,bool,bool,bool,bool,datetime[ms],f64,f64,datetime[ms],f64,f64,str,str,str,datetime[ms],str,datetime[ms],str,bool,bool,bool,bool,str,str,str,str,datetime[ms],f64,f64,datetime[ms],f64,f64,bool,f64,f64,f64,f64,str,bool,bool,bool,duration[ms],duration[ms],f64,f64
"""a001de""",null,"""N1VA""",2026-03-01 15:04:52,"""KRIC""",2026-03-01 17:29:12,"""KVJI""",false,true,false,false,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""df0_only""",false,false,false,null,null,null,null
"""a003a8""",null,"""N10PH""",2026-03-01 18:08:13,"""KBCT""",2026-03-01 20:52:44,"""KFXE""",false,true,false,false,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""df0_only""",false,false,false,null,null,null,null
"""a014b5""",null,"""N104MT""",2026-03-01 13:03:20,"""KBZN""",2026-03-01 16:56:40,"""KOPF""",false,true,false,false,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""df0_only""",false,false,false,null,null,null,null
"""a0180f""",null,"""N105J""",2026-03-01 12:31:49,"""KCAV""",2026-03-01 12:40:49,"""KIFA""",false,true,false,false,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""df0_only""",false,false,false,null,null,null,null
"""a01be7""",null,"""N106KH""",2026-03-01 20:09:38,"""MBPV""",2026-03-01 22:11:33,"""KSRQ""",false,true,false,false,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""df0_only""",false,false,false,null,null,null,null
…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…
"""adc204""",null,"""N986G""",2026-03-01 13:01:50,"""KFCI""",2026-03-01 13:24:50,"""KRIC""",false,true,false,false,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""df0_only""",false,false,false,null,null,null,null
"""adcca3""",null,"""N989AR""",2026-03-01 03:37:25,"""SBJH""",2026-03-01 14:26:44,"""KIAD""",false,true,false,false,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""df0_only""",false,false,false,null,null,null,null
"""adcca3""",null,"""N989AR""",2026-03-01 18:43:12,"""KIAD""",2026-03-01 19:26:40,"""KTEB""",false,true,false,false,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""df0_only""",false,false,false,null,null,null,null


In [9]:
df.filter(pl.col("icao") == "a003a8")

icao,callsign,registration,takeoff_time,takeoff_airport_ident,landing_time,landing_airport_ident,pia,ladd,military,interesting,first_message_time,first_lat,first_lon,last_message_time,last_lat,last_lon,icao_df1,callsign_df1,registration_df1,takeoff_time_df1,takeoff_airport_ident_df1,landing_time_df1,landing_airport_ident_df1,pia_df1,ladd_df1,military_df1,interesting_df1,aircraft_type,owner,aircraft_description,category,first_message_time_df1,first_lat_df1,first_lon_df1,last_message_time_df1,last_lat_df1,last_lon_df1,discard_df1,takeoff_airport_lat_df1,takeoff_airport_lon_df1,landing_airport_lat_df1,landing_airport_lon_df1,match_status,same_takeoff_airport_ident,same_landing_airport_ident,same_airport_ident,first_message_time_diff,last_message_time_diff,first_message_distance_to_takeoff_airport_km,last_message_distance_to_landing_airport_km
str,str,str,datetime[ms],str,datetime[ms],str,bool,bool,bool,bool,datetime[ms],f64,f64,datetime[ms],f64,f64,str,str,str,datetime[ms],str,datetime[ms],str,bool,bool,bool,bool,str,str,str,str,datetime[ms],f64,f64,datetime[ms],f64,f64,bool,f64,f64,f64,f64,str,bool,bool,bool,duration[ms],duration[ms],f64,f64
"""a003a8""",null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""a003a8""","""TWY116""","""N10PH""",2026-03-01 18:08:00,"""KBCT""",2026-03-01 18:59:00,"""MYEN""",false,true,false,false,"""C56X""","""BLUE SKY HARBOUR LLC""","""CESSNA 560XL Citation XLS""","""""",null,null,null,null,null,null,false,26.3785,-80.107697,24.594299,-76.820198,"""df1_only""",false,false,false,null,null,null,null
"""a003a8""",null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""a003a8""","""TWY116""","""N10PH""",2026-03-01 19:54:00,"""MYEN""",2026-03-01 20:52:00,"""KFXE""",false,true,false,false,"""C56X""","""BLUE SKY HARBOUR LLC""","""CESSNA 560XL Citation XLS""","""""",null,null,null,null,null,null,false,24.594299,-76.820198,26.1973,-80.1707,"""df1_only""",false,false,false,null,null,null,null
"""a003a8""",null,"""N10PH""",2026-03-01 18:08:13,"""KBCT""",2026-03-01 20:52:44,"""KFXE""",false,true,false,false,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"""df0_only""",false,false,false,null,null,null,null


In [ ]:
mismatches = df.filter(pl.col("match_status") != "both")
mismatch_summary = (
    mismatches.group_by("icao")
    .agg(
        (pl.col("match_status") == "df0_only").sum().alias("df0_only"),
        (pl.col("match_status") == "df1_only").sum().alias("df1_only"),
    )
    .sort("icao")
)

confirmed_gold_error_icaos = [
    "a1a4a7", "a1c2e3", "a1d684", "a3a1dc", "a3e348",
    "a63c0b", "a6ea99", "a875ab", "adccd0",
]
confirmed_both_error_icaos = ["a1e6b9"]

mismatch_classification = mismatch_summary.with_columns(
    pl.when(pl.col("icao").is_in(confirmed_both_error_icaos))
    .then(pl.lit("both"))
    .when(pl.col("icao").is_in(confirmed_gold_error_icaos))
    .then(pl.lit("df1_gold_incorrect"))
    .otherwise(pl.lit("adsbx_source_incorrect"))
    .alias("classification")
)
mismatch_classification

## Manual review of direct ADSB-X flight mismatches

This is the direct ADSB-X flight source (`get_flights(..., algorithm="adsbx")`), not the PlaneQuery flight algorithm using ADSB-X messages.

### Summary

- **6,244 true positives, 31 false positives, and 484 false negatives**
- **418 mismatched ICAOs / 515 mismatched rows**
- **Confirmed df1/gold errors:** 9 ICAOs
- **Both sources incorrect:** 1 ICAO
- **Direct ADSB-X source failures:** 408 ICAOs

### Mismatch error percentages

Percentages use the 418 mismatched ICAOs as the denominator.

| Test set Error (%) | Gold-Dataset Error (%) | Error in both (%) |
|---:|---:|---:|
| 97.61% | 2.15% | 0.24% |

The direct ADSB-X failures are overwhelmingly missing flights. A smaller group merges multiple SFDPS legs into a single long flight or truncates an international/low-coverage flight. The `mismatch_classification` dataframe above contains every ICAO and its category.

### Confirmed df1/gold errors

| ICAO | Reason |
|---|---|
| `a1a4a7` | ADSB-X correctly separates KBHM-KNEW-KBHM; df1 combines it. |
| `a1c2e3` | The real KAPF departure is around 20:33, not 18:19. |
| `a1d684` | ADSB-X correctly separates KACT-KTYR-KACT; df1 combines it. |
| `a3a1dc` | KRBD-KMKY departs around 21:10; df1 reuses 18:11. |
| `a3e348` | FA54-KHOU departs around 17:57; df1 reuses 15:07. |
| `a63c0b` | KENW-KBNA departs around 16:25; df1 reuses 12:48. |
| `a6ea99` | The real KGYY departure is around 16:36, not 15:30. |
| `a875ab` | KBKL-KLNS departs around 10:38; df1 reuses 06:12. |
| `adccd0` | KPHL-KHPN departs around 17:43; df1 reuses 14:03. |

### Both sources are incorrect

- `a1e6b9`: df1 gives the wrong KHOU-KDAL departure time, while the direct ADSB-X source also omits the KDAL-KSTL leg.

> **Comparison caveat:** `compare_airports=False` means airport-only disagreements can remain labeled `both`. Use `same_takeoff_airport_ident`, `same_landing_airport_ident`, and `same_airport_ident` when reviewing endpoint accuracy.